### DISCUSSION TITLE
Does Quartz.Extensions.DependencyInjection Work with Azure Function Apps?

### DISCUSSION THREAD

**Discussion by wjchristenson2** [quartznet/quartznet#1315] [Q&A]
I can't seem to get Quartz to work in Azure Function Apps using the Quartz.Extensions.DependencyInjection extensions. I am using Microsoft.Azure.Functions.Extensions to tap into the function app's startup. The IFunctionsHostBuilder has the IServiceCollection where I setup the Quartz service.  I've tried 2 methods of declaring the job/trigger.

```
[assembly: FunctionsStartup(typeof(WebAnalyticsItemsWorker.Startup))]
namespace WebAnalyticsItemsWorker
{
    public class Startup: FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddQuartz(q =>
            {
                q.UseMicrosoftDependencyInjectionJobFactory();

                //var jobKey = "extractMetricsIngestionJob";
                //q.AddJob<WebAnalyticsItemsWorker.Worker.ExtractMetricsIngestionJob>(opts => opts.WithIdentity(jobKey));
                //q.AddTrigger(opts => opts.ForJob(jobKey).WithIdentity($"{jobKey}-trigger").WithCronSchedule("0 0/1 * * * ?"));

                DateTimeOffset startTimeUtc = DateBuilder.EvenMinuteDate(DateTime.UtcNow);
                q.ScheduleJob<WebAnalyticsItemsWorker.Worker.ExtractMetricsIngestionJob>(triggerConfigurator =>
                    triggerConfigurator
                        .WithIdentity("extractMetricsIngestionJobTrigger")
                        .StartAt(startTimeUtc)
                        .WithCronSchedule("0 0/1 * * * ?"), //every 5-minutes at 0 seconds after the minute
                     jobConfigurator => jobConfigurator
                        .WithIdentity("extractMetricsIngestionJobConfig"));
            });
        }
    }
}
```

**Comment by lahma** [comment]
I haven't done any Azure Functions with Quartz so hard to tell if there should be better integration story here, but what is the problem?

It doesn't compile, it doesn't start or it doesn't run the jobs?

You are not adding the hosted service so you must manually get reference to `ISchedulerFactory` and do `var s = await schedulerFactory.GetScheduler(); await s.Start();`. This is basically what `AddQuartzHostedService` does.

**Reply by wjchristenson2** [reply]
> It doesn't compile, it doesn't start or it doesn't run the jobs?

The solution compiles, the application starts, however the job does not fire. I tried using the Quartz Hosted Service but it appears it is not compatible with Azure Function Apps as the app threw a lot of errors and wouldn't run.

> You are not adding the hosted service so you must manually get reference to ISchedulerFactory and do var s = await schedulerFactory.GetScheduler(); await s.Start();. This is basically what AddQuartzHostedService does.

How do you get a reference to the ISchedulerFactory to start it? I tried the following but I'm assuming it is simply creating a brand new scheduler and not using the Services.AddQuartz() scheduler. Manually creating the scheduler (below) with adding the triggers & jobs work, but I'd like to use the Quartz.Extensions.DependencyInjection so I can leverage Microsoft's DI.

```
            s_schedulerFactory = new StdSchedulerFactory();
            s_scheduler = await s_schedulerFactory.GetScheduler();
            await s_scheduler.Start();
```

**Reply by lahma** [reply]
You need to get a hold of `IServiceProvider`, basically `ISchedulerFactory factory = provider.GetRequiredService<ISchedulerFactory>();` or if functions support dependency injection via constructor, a constructor argument probably works.

**Reply by wjchristenson2** [reply]
> You need to get a hold of `IServiceProvider`, basically `ISchedulerFactory factory = provider.GetRequiredService<ISchedulerFactory>();` or if functions support dependency injection via constructor, a constructor argument probably works.

The issue with doing this is that I'm wanting to start the scheduler on Azure Function App Startup which is non-async, before functions are called, and DI can be used. I tried the following after the Quartz service was registered but it threw an exception when trying to call GetScheduler().

```
var schedulerFactory = builder.Services.BuildServiceProvider().GetRequiredService<ISchedulerFactory>();
var scheduler = WebAnalytics.Core.Threading.AsyncHelper.RunSync(schedulerFactory.GetScheduler());
```

> System.InvalidOperationException: 'Unable to resolve service for type 'Microsoft.Azure.WebJobs.Script.IFileLoggingStatusManager' while attempting to activate 'Microsoft.Azure.WebJobs.Script.Diagnostics.HostFileLoggerProvider'.'

Per [documentation](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection), it states to avoid using services at startup as additional function runtime dependencies are being loaded.

If the Quartz scheduler cannot auto-start, I'm guessing that I'll have to stick to using the vanilla Quartz implementation which I create a static instance on startup and kick it off. Unfortunately, I won't be able to use Microsoft DI with Quartz jobs this way.

**Reply by lahma** [reply]
Well without knowing what you actually are trying to achieve use case-wise, I think you could still change the job factory for the scheduler by doing this at the very beginning 

```csharp
IScheduler scheduler;  // get a reference
IServiceProvider serviceProvider; // from some injection?
scheduler.JobFactory = new MicrosoftDependencyInjectionJobFactory(serviceProvider, Options.Create(new QuartzOptions()))
```
To get similar scheduler building API that you've used to have in DI config, use `SchedulerBuilder`.

**Reply by wjchristenson2** [reply]
> Well without knowing what you actually are trying to achieve use case-wise

What I'm trying to achieve is wiring up Quartz in an Azure Function App via an FunctionsStartup class and use Microsoft DI w/ it. 

The issue is that I cannot use DI in the FunctionsStartup class to get a handle to the IServiceProvider or ISchedulerFactory. DI is only available by the function runtime when a function is called (after initial app startup). The only hack I can think of (which I wouldn't want to do) is to somehow manually trigger an initialize function after the app is started. DI can then be used to get a handle to the ISchedulerFactory and start its scheduler. Let me know if you have any other suggestions and I'll try it out.
